File: dong/java/src/main/java/P24SwapNodesInPairs.java
import bean.ListNode;

/**
 * 给定一个链表，两两交换其中相邻的节点，并返回交换后的链表。
 * <p>
 * 你不能只是单纯的改变节点内部的值，而是需要实际的进行节点交换。
 *
 * @author Lam
 * @ClassName P24SwapNodesInPairs
 * @date 2020/3/15
 */
public class P24SwapNodesInPairs {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode end = head == null ? null : head.next;
        while (end != null) {
            ListNode start = pre.next;
            ListNode next = end.next;
            end.next = start;
            start.next = next;
            pre.next = end;
            pre = start;
            end = next == null ? null : next.next;
        }

        return dummy.next;
    }
}

File: dong/go/p24swap_nodes_in_pairs.go
package main

func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	shim := new(ListNode)
	shim.Next = head
	pre := shim
	node := head
	next := head.Next

	for node != nil && next != nil {
		// pre - node - next
		// pre - next - node
		pre.Next = next
		node.Next = next.Next
		next.Next = node
		pre = node
		node = pre.Next
		if node != nil {
			next = node.Next
		}

	}

	return shim.Next
}
